python/lsst/ip/isr/brighterFatterKernel.py

```diff
@@ -0,0 +1,413 @@
+
+    ampKernels are the kernels for each amplifier in a detector, as
+    generated by having level == 'AMP'
+
+    detectorKernel is the kernel generated for a detector as a
+    whole, as generated by having level == 'DETECTOR'
+
+    makeDetectorKernelFromAmpwiseKernels is a method to generate the
+    kernel for a detector, constructed by averaging together the
+    ampwise kernels in the detector.
+
+    Parameters
+    ----------
+    level : `str`
+        Level the kernels will be generated for.
+    log : `lsst.log.Log`, optional
+        Log to write messages to.
+    **kwargs :
+        Parameters to pass to parent constructor.
+
+    """
+    _OBSTYPE = 'BFK'
+    _SCHEMA = 'Brighter-fatter kernel'
+    _VERSION = 1.0
+
+    def __init__(self, camera=None, level=None, **kwargs):
+        self.level = level
+
+        # Things inherited from the PTC
+        self.means = dict()
+        self.variances = dict()
+        self.rawXcorrs = dict()
+        self.badAmps = list()
+        self.shape = (8, 8)
+        self.gain = dict()
+        self.noise = dict()
+
+        # Things calculated from the PTC
+        self.meanXcorrs = dict()
+        self.valid = dict()
+
+        # Things that are used downstream
+        self.ampKernels = dict()
+        self.detKernels = dict()
+
+        if camera:
+            self.initFromCamera(camera, detectorId=kwargs.get('detectorId', None))
+
+        super().__init__(**kwargs)
+        self.requiredAttributes.update(['level', 'means', 'variances', 'rawXcorrs',
+                                        'badAmps', 'gain', 'noise', 'meanXcorrs', 'valid',
+                                        'ampKernels', 'detKernels'])
+
+    def updateMetadata(self, setDate=False, **kwargs):
+        """Update calibration metadata.
+
+        This calls the base class's method after ensuring the required
+        calibration keywords will be saved.
+
+        Parameters
+        ----------
+        setDate : `bool`, optional
+            Update the CALIBDATE fields in the metadata to the current
+            time. Defaults to False.
+        kwargs :
+            Other keyword parameters to set in the metadata.
+        """
+        kwargs['LEVEL'] = self.level
+        kwargs['KERNEL_DX'] = self.shape[0]
+        kwargs['KERNEL_DY'] = self.shape[1]
+
+        super().updateMetadata(setDate=setDate, **kwargs)
+
+    def initFromCamera(self, camera, detectorId=None):
+        """Initialize kernel structure from camera.
+
+        Parameters
+        ----------
+        camera : `lsst.afw.cameraGeom.Camera`
+            Camera to use to define geometry.
+        detectorId : `int`, optional
+            Index of the detector to generate.
+
+        Returns
+        -------
+        calib : `lsst.ip.isr.BrighterFatterKernel`
+            The initialized calibration.
+        """
+        self._instrument = camera.getName()
+
+        if self.level == 'AMP':
+            detector = camera[detectorId]
```

Comment:
`camera[detectorID]` won't fail if `detectorID` is `None`? (as it is initialized) Is it guaranteed that it will not be `None` when it reaches this point?

Reply:
The `detectorId` is required for the `level='AMP'` case.  I've added a `raise` if this isn't the case.